File: crates/core/src/roster.rs
//! WHO is loaded, once agents can be authored in the browser (increment 11):
//! the authored set folded out of the log, and the one function that makes the
//! running app agree with it. The ROUTES are `authoring.rs`.
//!
//! An authored agent is not a second kind of agent: it is the same `agent.md`
//! `public/agents/` serves, held as a fact in the event log instead of as a
//! file on a static host — which is why it survives a refresh (the log is
//! replayed at boot), why deleting one is just another fact (I10), and why it
//! is projected rather than stored twice (I8).
//!
//! PRECEDENCE, one rule — the Python `registry._agent_dirs` order plus one
//! step: built-ins, then `public/agents/`, then what this browser authored.
//! Last wins, so authoring `main` OVERRIDES the shipped `main` and deleting
//! that record reverts to the file. A live prompt edit is exactly this.

use kernel::{EventKind, EventLog, Status};

use crate::app::App;

/// An agent was written or replaced in this browser. Payload: `[name, text]`.
pub(crate) const AUTHORED: &str = "core.agent_authored";
/// An authored record was removed. Payload: the name.
pub(crate) const DELETED: &str = "core.agent_deleted";

/// Every agent this browser has authored and not deleted, as `(name, text)`
/// in the order they were first written — a fold over the log, like every
/// other view (I8).
pub(crate) fn authored(log: &EventLog) -> Vec<(String, String)> {
    let mut found: Vec<(String, String)> = Vec::new();
    for event in log.iter() {
        let EventKind::Custom { kind, payload_json } = &event.kind else {
            continue;
        };
        if kind == DELETED {
            if let Ok(name) = serde_json::from_str::<String>(payload_json) {
                found.retain(|(n, _)| *n != name);
            }
        } else if kind == AUTHORED {
            if let Ok((name, text)) = serde_json::from_str::<(String, String)>(payload_json) {
                match found.iter().position(|(n, _)| *n == name) {
                    Some(i) => found[i].1 = text,
                    None => found.push((name, text)),
                }
            }
        }
    }
    found
}

/// What THIS process authored — what a Worker hands back (`report_authored`).
pub fn authored_here(app: &App) -> Vec<(String, String)> {
    authored(&app.log)
}

/// An agent a SUB-AGENT wrote, adopted by the page. A Worker is its own Wasm
/// instance with its own event log, so `write_agent` called there records the
/// fact THERE and the page would never see it — the create-agent superagent
/// would report success and install nothing. Not the seam (I4): the host
/// reporting a fact, exactly like `report_agent` and `report_memory`, landing
/// as the same event the page's own form emits so there is one record and one
/// precedence rule. An identical repeat is dropped — a Worker re-reports its
/// whole authored set every turn.
pub fn report_authored(app: &mut App, name: &str, text: &str) {
    if authored(&app.log)
        .iter()
        .any(|(n, t)| n == name && t == text)
    {
        return;
    }
    app.append(EventKind::Custom {
        kind: AUTHORED.into(),
        payload_json: serde_json::to_string(&(name, text)).unwrap_or_default(),
    });
}

/// Make the running app agree with the authored set the log now holds.
///
/// DEFERRED WHILE A TURN IS IN FLIGHT. Swapping an agent's prompt between the
/// model call and the reply it is waiting for would assemble the rest of that
/// turn out of one file and the history of another's — the crossed-projection
/// class of bug increment 07 already produced once. `task` is `Some` exactly
/// from the utterance that starts a turn to the answer that ends it, so this
/// takes effect at a TURN BOUNDARY and the conversation is never disturbed.
pub(crate) fn reconcile(app: &mut App) {
    let want = authored(&app.log);
    if want == app.authored || app.agent.task.is_some() {
        return;
    }
    let files = crate::install::builtin_files()
        .into_iter()
        .chain(app.files.clone())
        .chain(want.clone());
    let (specs, problems) = agent::load_agents(files);
    let gone: Vec<String> = app
        .agents
        .iter()
        .filter(|s| !specs.iter().any(|n| n.name == s.name))
        .map(|s| s.name.clone())
        .collect();
    let fresh: Vec<String> = specs
        .iter()
        .filter(|s| !app.agents.iter().any(|o| o.name == s.name))
        .map(|s| s.name.clone())
        .collect();
    (app.authored, app.agents, app.agent_problems) = (want, specs, problems);
    rows(app, &gone, &fresh);
    let peers = app.agents.clone();
    let me = app.me.clone();
    if let Some(mine) = peers.iter().find(|s| s.name == me) {
        // The prompt, the toolbox and the space change; the paper's HISTORY is
        // untouched, which is what "the next turn uses the new one with the
        // conversation intact" means.
        agent::adopt_spec(&mut app.agent, mine, &peers);
    }
    let names: Vec<&str> = app.agents.iter().map(|s| s.name.as_str()).collect();
    app.append(EventKind::Custom {
        kind: "core.agents_loaded".into(),
        payload_json: serde_json::to_string(&names).unwrap_or_else(|_| "[]".into()),
    });
}

/// Board rows for agents that just appeared or went away. Only those:
/// re-registering everyone would reset running agents to `starting` and leave
/// them there — their Workers already reported and will not again.
fn rows(app: &mut App, gone: &[String], fresh: &[String]) {
    let now = app.ports.clock.now();
    let me = app.me.clone();
    for name in gone {
        app.board.forget(name);
    }
    for name in fresh {
        app.board.register(name, false, now);
        let status = match *name == me {
            true => Status::Waiting,
            false => Status::Starting,
        };
        app.board.set(name, status, "", now);
    }
}

/// The `write_agent` tool (increment 11's superagent runs on this one). An
/// ORDINARY tool: it appends the same fact the UI's form does, so an agent a
/// model wrote and one a person wrote are the same record, get the same
/// Worker, and are told apart on screen only by who wrote them (I9). The
/// capability it ends up with still comes from its SPACE and nowhere else.
pub(crate) fn write_agent(app: &mut App, args_json: &str) -> Result<String, String> {
    let value = serde_json::from_str::<serde_json::Value>(args_json).unwrap_or_default();
    let field = |k: &str| {
        value
            .get(k)
            .and_then(|v| v.as_str())
            .unwrap_or_default()
            .trim()
            .to_string()
    };
    let (name, prompt) = (field("name"), field("prompt"));
    if !agent::usable_agent_name(&name) {
        return Err(format!(
            "'{name}' cannot be an agent name — letters, digits, - and _ only. Call it as \
             write_agent({{\"name\": \"<name>\", \"description\": \"<one line>\", \
             \"prompt\": \"<the system prompt>\", \"tools\": \"\", \"space\": \"\"}})"
        ));
    }
    if prompt.is_empty() {
        return Err("no prompt given. An agent with no system prompt has no instructions at \
                    all; write the whole prompt as the 'prompt' argument."
            .into());
    }
    let tools: Vec<String> = field("tools").split(',').map(|t| t.trim().to_string()).collect();
    // A space that could never BE a space is dropped rather than written into
    // the file: `Space::named` would refuse it anyway, so keeping the string
    // would only put a capability line on the card that grants nothing. A real
    // name survives untouched.
    let space = agent::Space::named(&field("space")).map(|s| s.name).unwrap_or_default();
    let spec = agent::new_spec(&name, &field("description"), &unescaped(&prompt), &tools, &space);
    let text = agent::render_agent_file(&spec);
    app.append(EventKind::Custom {
        kind: AUTHORED.into(),
        payload_json: serde_json::to_string(&(name.clone(), text)).unwrap_or_default(),
    });
    Ok(format!(
        "Wrote {name}. It is installed in this browser as soon as this turn ends — it has its \
         own Worker and its own conversation, and it is listed beside the shipped agents. \
         Tell the user it exists and what to ask it."
    ))
}

/// A prompt a model wrote with its newlines still escaped. Small local models
/// double-escape a multi-line string inside a one-line call often enough that
/// the agents they write arrive as one 400-character paragraph.
///
/// ponytail: only when there is no real newline to lose — a prompt that
/// already has line breaks is passed through untouched, so this can only fix a
/// prompt that has none.
fn unescaped(prompt: &str) -> String {
    match prompt.contains('\n') {
        true => prompt.to_string(),
        false => prompt.replace("\\n", "\n").replace("\\t", "\t"),
    }
}

File: packages/core/src/roster.js
/**
 * THE TWO TOOLS THAT ACT ON THE ROSTER: one writes an agent into this browser,
 * the other sets one working. Each hands capability to something that is not
 * this turn, which is why their descriptions are the longest in the build.
 *
 * `write_agent` IS AN ORDINARY TOOL (I9). It appends exactly the facts the
 * `/agents` route appends, so an agent a model wrote and one a person wrote are
 * the same record, boot the same way, and are told apart on screen only by who
 * wrote them.
 *
 * WHAT THE RUST DEFERRED AND THIS DOES NOT. There, an authored agent was
 * unreachable until the turn ended, because `reconcile` swapped the RUNNING
 * agent's prompt and could not do that mid-turn. Here a sub-agent boots FRESH
 * in its own Worker and reads the log for its file, so there is nothing to swap
 * and a `spawn_agent` naming an agent written this turn simply works. The
 * ordering rule went with the mechanism that needed it.
 *
 * THE `space` SENTENCE IS THE ONE THE RUST GOT WRONG TWICE — it claimed naming
 * a space also granted the workspace tools, and then its hand-kept list of them
 * went stale. In this build a space grants the space faculty and nothing else;
 * the workspace tools are in the catalogue for every agent to name. There is no
 * list here to go stale.
 * @module
 */

import { arg, loadAgents, tool } from '@harness/agent'

import { AGENT_AUTHORED, AGENT_MODULE } from './agents.js'
import { answered, nameArg } from './runner.js'

/** @typedef {import('./app.js').App} App */
/** @typedef {import('./app.js').ToolRun} ToolRun */

/** The descriptors, so a model that may call these is TOLD how. */
export const ROSTER_TOOLS = [
  tool({
    name: 'write_agent',
    description: 'create or replace an agent in this browser. It gets its own Worker, its own conversation and its own tools, and it is listed beside the shipped agents. "tools" is a comma-separated list of tool names, and empty means every tool this build offers',
    args: [
      arg('name', 'string', 'what it is called: letters, digits, - and _ only'),
      arg('description', 'string', 'one line saying what it is for — this is the line other agents read'),
      arg('prompt', 'string', 'the whole system prompt, as the agent will be told it'),
      arg('tools', 'string', 'the tools it may call, comma-separated; empty means all of them', { required: false }),
      arg('space', 'string', 'the shared space it works in, if any', { required: false }),
    ],
    mutates: true,
    needs: 'emit',
  }),
  tool({
    name: 'spawn_agent',
    description: 'hand a goal to an agent that already exists: it works on it in its own Worker, with its own tools and its own conversation, and its answer comes back as this call\'s result. It creates nothing and can lend nothing — list_agents says which agents exist, and write_agent is what authors one',
    args: [
      arg('agent', 'string', 'the agent to set working, as list_agents spells it'),
      arg('goal', 'string', 'the whole task, in one string — it cannot see this conversation'),
    ],
    needs: 'agents',
  }),
]

/**
 * The runners, over the App that holds the log and the port.
 * @param {App} app @returns {Record<string, ToolRun>}
 */
export function rosterTools(app) {
  return {
    write_agent: answered('write_agent', async (args) => wrote(app, args)),
    spawn_agent: answered('spawn_agent', async (args, opts) => sent(app, args, opts.signal)),
  }
}

/** Letters, digits, `-` and `_`. The name keys the roster, addresses a message and names a segment stream, so anything else is a name three subsystems would spell differently. */
const USABLE = /^[A-Za-z0-9_-]+$/

/**
 * WRITE ONE AGENT FILE, AS THE FACTS THE ROSTER IS FOLDED FROM.
 *
 * The file is rendered and then READ BACK before anything is recorded: a spec
 * this build would refuse must be refused HERE, where the model can fix it,
 * rather than at the next boot where it becomes a row in the agents pane.
 * @param {App} app @param {Record<string, unknown>} args
 */
function wrote(app, args) {
  const name = nameArg(args, 'name')
  const prompt = unescaped(String(args['prompt'] ?? ''))
  if (!USABLE.test(name)) {
    return { ok: false, output: `"${name}" cannot be an agent name — letters, digits, - and _ only. Call it as write_agent({"name": "<name>", "description": "<one line>", "prompt": "<the system prompt>"}).` }
  }
  if (prompt.trim() === '') {
    return { ok: false, output: 'No prompt given, and an agent with no system prompt has no instructions at all. Write the whole prompt as the "prompt" argument.' }
  }
  const path = `${name}/agent.md`
  const text = agentFile(name, args, prompt)
  const refusal = loadAgents([{ path, text }]).refusals[0]
  if (refusal) return { ok: false, output: `That would not load: ${refusal.message}` }
  const at = app.ports.clock.now()
  app.log.append({ type: 'custom', kind: AGENT_AUTHORED, payload: { name, path, text } }, at)
  app.log.append({ type: 'module_installed', module: `${AGENT_MODULE}${name}`, version: '1' }, at)
  return { ok: true, output: `Wrote ${name}. It is installed in this browser now — spawn_agent can reach it, it is listed beside the shipped agents, and it survives a reload. Tell the person it exists and what to ask it.` }
}

/**
 * THE FILE ITSELF: frontmatter for what the machine reads, the body for what
 * the model reads. Rendered rather than assembled from a template string
 * elsewhere, because `parseAgentFile` is the only reader and the two have to
 * agree about one format.
 * @param {string} name @param {Record<string, unknown>} args @param {string} prompt
 */
function agentFile(name, args, prompt) {
  const listed = nameArg(args, 'tools').split(',').map((t) => t.trim()).filter((t) => t !== '')
  const space = nameArg(args, 'space')
  const lines = [
    '---',
    `name: ${name}`,
    `description: ${String(args['description'] ?? '').replace(/\n/g, ' ').trim()}`,
    ...(listed.length > 0 ? ['tools:', ...listed.map((t) => `  - ${t}`)] : []),
    ...(space === '' ? [] : [`space: ${space}`]),
    '---',
    '',
  ]
  return `${lines.join('\n')}${prompt.trim()}\n`
}

/**
 * A prompt a model wrote with its newlines still escaped. Small local models
 * double-escape a multi-line string inside a one-line call often enough that
 * the agents they write arrive as one 400-character paragraph. Only where there
 * is no real newline to lose: a prompt that already has line breaks is passed
 * through untouched, so this can only fix a prompt that has none.
 * @param {string} prompt @returns {string}
 */
function unescaped(prompt) {
  return prompt.includes('\n') ? prompt : prompt.replace(/\\n/g, '\n').replace(/\\t/g, '\t')
}

/**
 * HAND A GOAL TO AN AGENT THAT EXISTS. An empty goal is refused here and never
 * delivered: a sub-agent handed one answers it regardless, which costs a whole
 * turn of somebody else's loop to learn nothing.
 *
 * The port's own refusal is the result on the way back — it names which agents
 * this build can run, and a model that reads it can pick another (I15).
 * @param {App} app @param {Record<string, unknown>} args @param {AbortSignal} signal
 */
async function sent(app, args, signal) {
  const agent = nameArg(args, 'agent')
  const goal = String(args['goal'] ?? '').trim()
  if (agent === '' || goal === '') {
    const missing = agent === '' ? 'No agent named — list_agents says which exist' : 'No goal given, and an agent handed an empty goal answers it anyway'
    return { ok: false, output: `${missing}. Call it as spawn_agent({"agent": "<one that already exists>", "goal": "<the whole task, in one string>"}).` }
  }
  return { ok: true, output: await app.ports.agents.delegate(agent, goal, { signal }) }
}
